subsys/net/lib/sockets/sockets.c

```diff
@@ -849,3 +873,31 @@ int zsock_setsockopt(int sock, int level, int optname,
 	errno = ENOPROTOOPT;
 	return -1;
 }
+
```

Comment:
Thanks for looking into this. So, the conceptual problem here is that we now implement our own POSIX API with POSIX names, and names like recv() now belong to POSIX API. socket.c is supposed to *implement* those calls. If implementation calls recv(), which it's supposed to implement, we get (conceptual) infinite loop.

I don't know if it's real or not, but at least I don't think that there should be a patch doing that right in this PR. Once that tested to work, might go as a temporary solution.

Otherwise, I told that trying to implement offloading on top of the native socket API using defines (or inlines) is cheating. The proper layering should be:

~~~
(optionally) socket -> zsock_socket() --> net_ctx
                                      |
                                      |-> offload
                                      |
                                      \-> TLS
~~~

Because now that becomes

~~~
(optionally) socket -> zsock_socket() -> fdtable -> net_ctx
~~~

And until you guys update to use fdtable, read() and write() will be out of reach for your (pseudo)sockets. And they are "pseudo" because as far as I can tell, syscalls (i.e. proper userspace vs kernel seperation and implementation) never worked for your sockets either.